### ISSUE TITLE
[EngSys] Validate Changelog before merge instead of during release

### ISSUE THREAD

**Issue by xirzec** [Azure/azure-sdk-tools#1878]
After validating a release I was met with this error:

##[error]The changelog entry has the following sections with no content (Bugs Fixed, Breaking Changes, Other Changes). Please ensure to either remove the empty sections or add content to the section.

We should either move this check as-is into the PR CI gates or if we want to be really fancy, make it happen when we go from Unreleased - > a real date.

**Comment by weshaggard** [comment]
@xirzec did you run the Prepare-Release script? It would have caught this issue and other potential things like unapproved apis.

That said @praveenkuttappan and I chatted and we are also going to look to start enforcing more of these in the build process when we see the date stamped into the changelog.

**Comment by praveenkuttappan** [comment]
@xirzec and I chatted now and found that prepare release script was not run in this case.

**Comment by weshaggard** [comment]
Just to clarify @heaths  Prepare-Release did error on the empty sections correct? If not that is a bug that should be fixed. As for the request to have it strip empty sections that is reasonable and we can probably add that, issue filed https://github.com/Azure/azure-sdk-tools/issues/1895.

**Comment by heaths** [comment]
It did not. That said, I ran it mid-last week. Not sure if there were changes since then. My PR (it's resulting output) was just sitting, waiting for release day.

**Comment by weshaggard** [comment]
That change should have already been in place for a while, I'll have @chidozieononiwu double check when he works on issue I opened.

**Comment by weshaggard** [comment]
@chidozieononiwu use this issue to check whether or not we are actually running the correct changelog validation steps during prepare-prerelease.

**Comment by witemple-msft** [comment]
Ran into this exact issue as well for @azure/ai-form-recognizer. I had an empty "Breaking Changes" section in the log. I _did_ run the Prepare-Release script and it did not produce an error, but the pipeline validation failed after approving the release.

For what it's worth, I think all validation steps that might result in a new commit to the repo to resolve them have to be run before approval, since I can't approve the pipelines until I've already gone through the painstaking process of validating the artifact and I have to assume that it will publish successfully once I approve it. If it fails, I have to fix the package and generate and test a new candidate artifact.

**Comment by heaths** [comment]
It would be helpful if,

1. The script just removed empty sections. It's already making changes to the file e.g., add the release date.
2. The release pipeline validate empty sections during the build jobs so we don't waste time validating the signed bits only to find out there was a doc error.

**Comment by weshaggard** [comment]
Yes the goal is to have Prepare-Release do all the work and validation that would happen during release and also makes sense for it to do any last edits that are needed before a release. 

As for validation during the build that is a little more tricky because we don't always know that we are going to release or not but we are thinking about adding more build step validation when we see that you have a release date stamped into your changelog which tells us you plan to release soon.

**Comment by heaths** [comment]
How is manually running, say, "net - keyvault" different from it running nightly? There's a condition there (don't feel like looking now, so the aforementioned question wasn't pointed 🙂) you could probably use to determine when you're running the pipeline for a release vs. scheduled (and maybe that is literally it - IIRC, there's an AzPipelines variable that gives you a reason).

**Comment by heaths** [comment]
From [docs](https://docs.microsoft.com/en-us/azure/devops/pipelines/build/variables?view=azure-devops&tabs=yaml#build-variables-devops-services):

`Build.Reason`

There's probably more to the condition to be reliable than that, but one idea to validate earlier but conditionally.

**Comment by weshaggard** [comment]
When Build.Reason is manual it might mean you want to do a release but it might not people do manual builds without the intent of releasing. So it isn't the best variable to use. We could introduce another variable but I think the best way we have currently is by checking the release date stamped in the changelog. If you have that then any build we do should start failing if you aren't meeting all the release criteria (i.e. changelog verification, api approval etc)?

**Comment by heaths** [comment]
Releasing on the actual release date won't always work: many start it early or, like in this case because of the issues this caused, I started late. For various reasons, that's also not unheard of and I've not seen anyone go back and fix their changelog dates.

The _archtype-sdk-client.yml_ currently just uses:

```yaml
  - ${{if and(ne(variables['Build.Reason'], 'PullRequest'), eq(variables['System.TeamProject'], 'internal'))}}:
```

If the release stage would run even in that case, why not check the changelog as well?

**Comment by weshaggard** [comment]
The trouble is that if we run it in that case and there are multiple sdk's in a given pipeline and you only plan to release a subset of them then the other ones that aren't planning to release and aren't in a valid state will block you from releasing the ones that are in a valid state. 

As for the date it doesn't matter what the date is it only matters there is a date as opposed to unreleased and if there is a date then we can assume that you plan to release soon and add the extra validation for that library.

**Comment by heaths** [comment]
How about not adding them in the first place and just check that all header names are standard (what's documented)? Start from clean.

**Comment by weshaggard** [comment]
We definitely will be adding those checks in Prepare-Release and we will also start cleaning out empty sections. I thought it was already enabled there but based on the feedback there must be a bug there.
